### PR TITLE
[flang] Implement the UNPACK intrinsic

### PR DESCRIPTION
**Pull request by psteinfeld**
I implemented the lowering and augmented the existing runtime support by
supporting a scalar version of the "field" argument.  I added tests for
lowering and a runtime test with a scalar value for the "field" argument.

### CODE REVIEW

File: flang/lib/Lower/IntrinsicCall.cpp [hunk 1]
```diff
@@ -2794,6 +2799,39 @@ static mlir::Value createExtremumCompare(mlir::Location loc,
...
+fir::ExtendedValue
+IntrinsicLibrary::genUnpack(mlir::Type resultType,
+                            llvm::ArrayRef<fir::ExtendedValue> args) {
+  assert(args.size() == 3);
+
+  // Handle required vector argument
+  auto vector = builder.createBox(loc, args[0]);
+  fir::BoxValue vectorTmp = builder.createBox(loc, args[0]);
+  auto vectorRank = vectorTmp.rank();
+  assert(vectorRank == 1);
```
Comment (schweitzpgi) [inline flang/lib/Lower/IntrinsicCall.cpp:2812]: Shouldn't the rank of the vector argument be checked in semantics?

File: flang/test/Lower/transformational-intrinsics.f90 [hunk 2]
```diff
@@ -124,4 +124,18 @@ subroutine in_elem_expr(x, y, z)
...
+subroutine unpack_test()
+  integer, dimension(3) :: vector
+  integer, dimension (3,3) :: field
+
+  logical, dimension(3,3) :: mask
+  integer, dimension(3,3) :: result
+  result = unpack(vector, mask, field)
+  ! CHECK: fir.call @_FortranAUnpack(%{{.*}}, %{{.*}}, %{{.*}}, %{{.*}}, %{{.*}}, %{{.*}}) : (!fir.ref<!fir.box<none>>, !fir.box<none>, !fir.box<none>, !fir.box<none>, !fir.ref<i8>, i32) -> none
+  result = unpack(vector, mask, 343)
+  ! CHECK: fir.call @_FortranAUnpack(%{{.*}}, %{{.*}}, %{{.*}}, %{{.*}}, %{{.*}}, %{{.*}}) : (!fir.ref<!fir.box<none>>, !fir.box<none>, !fir.box<none>, !fir.box<none>, !fir.ref<i8>, i32) -> none
```
Comment (schweitzpgi) [inline flang/test/Lower/transformational-intrinsics.f90:138]: Shouldn't these do at least some cursory CHECKs on the arguments?

**Review by schweitzpgi — COMMENTED** [review]
Nice work, Pete.

File: flang/test/Lower/transformational-intrinsics.f90 (hunk 2 above)
Comment (mleair) [reply flang/test/Lower/transformational-intrinsics.f90:138]: Take a look at how some of the other tests are done, like reshape. As Eric said, you should check the arguments. Also, add a check on the call to fir.freemem at the end.

File: flang/lib/Lower/IntrinsicCall.cpp (hunk 1 above)
Comment (psteinfeld) [reply flang/lib/Lower/IntrinsicCall.cpp:2812]: You're correct.  I'll remove this

File: flang/test/Lower/transformational-intrinsics.f90 (hunk 2 above)
Comment (psteinfeld) [reply flang/test/Lower/transformational-intrinsics.f90:138]: I wasn't sure about the level of checking needed for argument passing.  The other tests don't seem consistent about this.  I'll add some checks.

Comment (schweitzpgi) [reply flang/test/Lower/transformational-intrinsics.f90:138]: I think we'll want to write more restrictive tests before we try to upstream to help prevent helpful rerouting suggestions sending us over cliffs.

Comment (psteinfeld) [reply flang/test/Lower/transformational-intrinsics.f90:138]: I'll add checks for the arguments.

File: flang/test/Lower/transformational-intrinsics.f90 [hunk 3]
```diff
@@ -124,4 +124,59 @@ subroutine in_elem_expr(x, y, z)
...
+  ! CHECK-DAG: %[[a44:.*]] = fir.shape %{{.*}}, %{{.*}} : (index, index) -> !fir.shape<2>
+  ! CHECK-DAG: %[[a45:.*]] = fir.embox %[[a43]](%[[a44]]) : (!fir.heap<!fir.array<?x?xi32>>, !fir.shape<2>) -> !fir.box<!fir.heap<!fir.array<?x?xi32>>>
+  ! CHECK-DAG: fir.store %[[a45]] to %[[a0]] : !fir.ref<!fir.box<!fir.heap<!fir.array<?x?xi32>>>>
+  ! CHECK-DAG: %[[a47:.*]] = fir.convert %[[a0]] : (!fir.ref<!fir.box<!fir.heap<!fir.array<?x?xi32>>>>) -> !fir.ref<!fir.box<none>>
+  ! CHECK-DAG: %[[a48:.*]] = fir.convert %[[a39]] : (!fir.box<!fir.array<3xi32>>) -> !fir.box<none>
+  ! CHECK-DAG: %[[a49:.*]] = fir.convert %[[a41]] : (!fir.box<!fir.array<3x3x!fir.logical<4>>>) -> !fir.box<none>
+  ! CHECK-DAG: %[[a50:.*]] = fir.convert %[[a42]] : (!fir.box<i32>) -> !fir.box<none>
+  result = unpack(vector, mask, 343)
+  ! CHECK: fir.call @_FortranAUnpack(%[[a47]], %[[a48]], %[[a49]], %[[a50]], %{{.*}}, %{{.*}}) : (!fir.ref<!fir.box<none>>, !fir.box<none>, !fir.box<none>, !fir.box<none>, !fir.ref<i8>, i32) -> none
+  ! CHECK: fir.freemem %{{.*}} : !fir.heap<!fir.array<?x?xi32>>
```
Comment (mleair) [inline flang/test/Lower/transformational-intrinsics.f90:179]: I suggest including the fir argument for freemen. I believe this just adds two more CHECK-DAG lines (at least that's been the case on the regression tests I have added)

Comment (mleair) [reply flang/test/Lower/transformational-intrinsics.f90:179]: For example, reshape_test:

! CHECK-DAG:  %[[a15:.*]] = fir.load %[[a0]] : !fir.ref<!fir.box<!fir.heap<!fir.array<?x?xi32>>>>
! CHECK-DAG:  %[[a18:.*]] = fir.box_addr %[[a15]] : (!fir.box<!fir.heap<!fir.array<?x?xi32>>>) -> !fir.heap<!fir.array<?x?xi32>>
! CHECK-DAG:  fir.freemem %[[a18]] : !fir.heap<!fir.array<?x?xi32>>

Comment (psteinfeld) [reply flang/test/Lower/transformational-intrinsics.f90]: Thanks, Mark.  I just pushed another commit with this change.

File: flang/test/Lower/transformational-intrinsics.f90 [hunk 4]
```diff
@@ -124,4 +124,61 @@ subroutine in_elem_expr(x, y, z)
...
+  ! CHECK-DAG: %[[a44:.*]] = fir.shape %{{.*}}, %{{.*}} : (index, index) -> !fir.shape<2>
+  ! CHECK-DAG: %[[a45:.*]] = fir.embox %[[a43]](%[[a44]]) : (!fir.heap<!fir.array<?x?xi32>>, !fir.shape<2>) -> !fir.box<!fir.heap<!fir.array<?x?xi32>>>
+  ! CHECK-DAG: fir.store %[[a45]] to %[[a0]] : !fir.ref<!fir.box<!fir.heap<!fir.array<?x?xi32>>>>
+  ! CHECK-DAG: %[[a47:.*]] = fir.convert %[[a0]] : (!fir.ref<!fir.box<!fir.heap<!fir.array<?x?xi32>>>>) -> !fir.ref<!fir.box<none>>
+  ! CHECK-DAG: %[[a48:.*]] = fir.convert %[[a39]] : (!fir.box<!fir.array<3xi32>>) -> !fir.box<none>
+  ! CHECK-DAG: %[[a49:.*]] = fir.convert %[[a41]] : (!fir.box<!fir.array<3x3x!fir.logical<4>>>) -> !fir.box<none>
+  ! CHECK-DAG: %[[a50:.*]] = fir.convert %[[a42]] : (!fir.box<i32>) -> !fir.box<none>
+  result = unpack(vector, mask, 343)
+  ! CHECK: fir.call @_FortranAUnpack(%[[a47]], %[[a48]], %[[a49]], %[[a50]], %{{.*}}, %{{.*}}) : (!fir.ref<!fir.box<none>>, !fir.box<none>, !fir.box<none>, !fir.box<none>, !fir.ref<i8>, i32) -> none
+  ! CHECK-DAG: %[[a53:.*]] = fir.load %0 : !fir.ref<!fir.box<!fir.heap<!fir.array<?x?xi32>>>>
```
Comment (mleair) [inline flang/test/Lower/transformational-intrinsics.f90:179]: Need %[a0] instead of %0.

Comment (psteinfeld) [reply flang/test/Lower/transformational-intrinsics.f90]: Ay yi yi.  Thanks, Mark.  Just pushed a fix.

File: flang/runtime/transformational.cpp [hunk 5]
```diff
@@ -521,9 +521,10 @@ void RTNAME(Unpack)(Descriptor &result, const Descriptor &vector,
     }
     result.IncrementSubscripts(resultAt);
     mask.IncrementSubscripts(maskAt);
-    field.IncrementSubscripts(fieldAt);
+    if (field.rank() != 0) { // scalar "field" argument
+      field.IncrementSubscripts(fieldAt);
+    }
```
Comment (mleair) [inline flang/runtime/transformational.cpp:526]: Did you already add this change to llvm-project? 

@schweitzpgi is it better to put this change in a separate PR if it is already added to llvm-project? 

Or can this change wait until fir-dev is upstreamed to llvm-project?

File: flang/unittests/RuntimeGTest/Transformational.cpp [hunk 6]
```diff
@@ -200,4 +200,20 @@ TEST(Transformational, Unpack) {
     EXPECT_EQ(*result.ZeroBasedIndexedElement<std::int32_t>(j), expect[j]);
   }
   result.Destroy();
+
+  // Test for scalar value of the "field" argument
+  auto scalarField{MakeArray<TypeCategory::Integer, 4>(
```
Comment (mleair) [inline flang/unittests/RuntimeGTest/Transformational.cpp:205]: Same question as flang/runtime/transformational.cpp  (@psteinfeld and @schweitzpgi)

File: flang/runtime/transformational.cpp (hunk 5 above)
Comment (schweitzpgi) [reply flang/runtime/transformational.cpp:526]: Normally, we just cherry pick over any code from llvm-project, keeping the original change as intact as is possible. For example, we wouldn't squash the cherry-picked changes with a bunch of other changes. That helps both git and humans figure out what is happening. :)

Caveat: it may be that only part of an llvm-project commit can be cherry-picked. That should be fine as well. What we are really trying to avoid are make-work sorts of conflicts.

Comment (psteinfeld) [reply flang/runtime/transformational.cpp:526]: But in this case, I don't see this change in the "main" branch on llvm-project.

File: flang/unittests/RuntimeGTest/Transformational.cpp (hunk 6 above)
Comment (psteinfeld) [reply flang/unittests/RuntimeGTest/Transformational.cpp:205]: > 
> 
> Same question as flang/runtime/transformational.cpp (@psteinfeld and @schweitzpgi)

I don't see any similar change on the "main" branch of the llvm-project.

Comment (mleair) [reply flang/unittests/RuntimeGTest/Transformational.cpp:205]: OK. So, are you planning on just keeping these changes in fir-dev for now? @schweitzpgi can confirm this...but I assume they would then go into "main"of llvm-project when fir-dev is upstreamed.

File: flang/runtime/transformational.cpp (hunk 5 above)
Comment (schweitzpgi) [reply flang/runtime/transformational.cpp:526]: Should it be? Or this is completely new work that hasn't been merged into llvm-project yet?

Either way, we can merge it into fir-dev. We just want to keep these changes "in synch" somehow, right?

Comment (psteinfeld) [reply flang/runtime/transformational.cpp:526]: The main reason for this change is to improve our results on the tests that Steve Scalpone is running, which I assume are the result of building the compiler from fir-dev.

I was not personally planning to move these changes into llvm-project.

**Review by mleair — APPROVED** [review]
Looks good. Just a reminder, squash your commits before merging in the changes.
